Using index mappers instead of own created storage for states of filtering

wszymanski opened the pull request:
### Context
PR https://github.com/handsontable/handsontable/pull/7276 introduced a new kind of Index Map. I'm using it here for purpose of storing and managing states of filtering for particular columns. The `Filters` plugin has been adjusted to use the only Index Map (all unnecessary parts of code has been removed). What's more:

- I removed `clearConditions` method of `ConditionCollection` element. There was a function that cleared conditions at the specified column index but without clearing stack order (stored in separate array). Since the order of columns is stored inside the Index Mapper there is no way to threaten it separately. Thus, some parts of the code have been changed. Hooks `beforeClear` and `afterClear` have been removed while `afterRemove` hook has been introduced.
- I changed `isMatchInConditions` method of `ConditionCollection` element. It will need an argument. Function without an argument hasn't been used internally.
- I introduced an extra method - named `clearValue` - for newly created Index Map within https://github.com/handsontable/handsontable/pull/7276.
- The internal "hiddenRowsCache" map was removed as it turned out that is not necessary at all. The UI components' hidden state can be calculated on the fly while showing the Dropdown
menu.
 - The Filters UI components previously used the `stateSaver` mixing which causes the components state misalignment after column reordering. The mixin is removed and the state management from now on is implemented directly into the UI Base component.

### How has this been tested?
<!--- Please describe in detail how you tested your changes (doesn't apply to translations). -->
Manual tests for alter actions.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #6830
2. #6397

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation.

budnix commented on src/translations/maps/linkedPhysicalIndexToValueMap.js:
```diff
@@ -75,6 +76,22 @@ class LinkedPhysicalIndexToValueMap extends IndexMap {
+   *
+   * @param {number} physicalIndex Physical index.
+   */
+  clearValue(physicalIndex) {
```
I haven't found any unit test for that method. Could I ask you to cover that?

wszymanski replied:
The `setValueAtIndex` already removes a value from the mentioned array.

https://github.com/handsontable/handsontable/blob/8ee91484fe1efa000c577faadbcdc11fac0d8f87/src/translations/maps/linkedPhysicalIndexToValueMap.js#L54-L60

However, the method does extra execution of `initValueOrFn` function or pass initial value to the `setValueAtIndex`.

I've used `clearValue` name as it works very similar to the current `clear` method.

wszymanski replied:
Added test within https://github.com/handsontable/handsontable/pull/7319/commits/c4091ccc7d255d07298631285aa5b3c8aa563b7d.

budnix reviewed (changes requested):
You changed fewer lines of code than I originally assumed 👍  Good Job! I left some observations and comments in the source code.

wszymanski commented on src/plugins/filters/filters.js:
```diff
@@ -194,12 +197,13 @@ class Filters extends BasePlugin {
     }
     if (!this.conditionCollection) {
-      this.conditionCollection = new ConditionCollection();
+      this.conditionCollection = new ConditionCollection(this.sortingStates);
```
That's the right direction, but there is a problem. Created index mappers within the mentioned PR are also registered. I'm afraid, that there should be no registration for one use case.

https://github.com/handsontable/handsontable/blob/46c62ae1d0593fe2b4f49f5f3bcf6dbf1830cb9b/src/plugins/filters/conditionUpdateObserver.js#L139

This temporary collection isn't registered to the IndexMapper. WDYT?

budnix replied:
Maybe we could add an option to the `ConditionCollection` class wherein the constructor the map registration will be differentiated according to the existence of the option.

So the constructor signature could look something like:
```
new ConditionCollection(hot, { trackIndexes: true });
```
PS: I believe that you would find a better name for the `trackIndexes` option.

wszymanski commented on src/plugins/filters/filters.js:
```diff
@@ -139,6 +139,9 @@ class Filters extends BasePlugin {
 
     this.filtersRowsMap = this.hot.rowIndexMapper.registerMap(this.pluginName, new TrimmingMap());
     this.dropdownMenuPlugin = this.hot.getPlugin('dropdownMenu');
+    this.sortingStates = this.hot.columnIndexMapper.registerMap(
```
Changed within https://github.com/handsontable/handsontable/pull/7319/commits/c7eb3b9c7e3acfb339fcbdf836c998bf80a4e108.

wszymanski commented on src/plugins/filters/conditionCollection.js:
```diff
@@ -166,20 +153,13 @@ class ConditionCollection {
-      const operation = this.columnTypes[column];
-
-      result.push({
+    return arrayReduce(this.filteringStates.getEntries(), (allConditions, [column, { operation, conditions }]) => {
```
Changed within https://github.com/handsontable/handsontable/pull/7319/commits/c7eb3b9c7e3acfb339fcbdf836c998bf80a4e108.

wszymanski commented:
> You changed fewer lines of code than I originally assumed 👍 Good Job! I left some observations and comments in the source code.
> 
> Additionally, I've found that the TS definition wasn't updated yet. Mostly about `ConditionCollection` class.

Thanks. Documentation should be also updated.

budnix commented:
I think that I've found a bug. When you filter columns using the "filter_by_value" component the data within that component don't react after changing the conditions from the previous columns.

Before the PR:
![beforePR](https://user-images.githubusercontent.com/571316/96458586-61dc9980-1221-11eb-8c42-9cb37baf0777.gif)

Within the PR
![afterPR](https://user-images.githubusercontent.com/571316/96458635-6e60f200-1221-11eb-9a48-89c36ea47c44.gif)

wszymanski commented:
Within this PR I removed `clearConditions` method of ConditionCollection element. There was a function that cleared conditions at the specified column index but **without clearing stack order** (stored in a separate array). Since the order of columns is stored inside the Index Mapper there is no way to threaten it separately in a clean way.

Bug found by @budnix was related to the fact that after refactor, the drop-down menu removed (not only "temporary cleared" value) all conditions for a particular column (they may be overwritten). Thus, the filtered column was completely removed from the queue of filtering.

https://github.com/handsontable/handsontable/blob/9375365c5f566c6e829ce252069de68661285768/src/plugins/filters/filters.js#L613-L621

In the next step, it could be added to the queue but **not in the previous position** but at the last position.

https://github.com/handsontable/handsontable/blob/9375365c5f566c6e829ce252069de68661285768/src/plugins/filters/filters.js#L623-L633

I've added an extra argument to the `setValueAtIndex` of the newly created index map.

https://github.com/handsontable/handsontable/blob/63c22ace8677646d7c74b6cc3f96c6805ed9412e/src/translations/maps/linkedPhysicalIndexToValueMap.js#L55

It should fix the problem. I've prepared also changes similar to this suggested by @budnix within [a comment](https://github.com/handsontable/handsontable/pull/7319#discussion_r508262099). A way how we create states manager for the `ColumnSorting` and `Filters` plugins is very similar from now.

PS. I've tried to add one more extra argument to the `setValueAtIndex` with quite unnatural default. I was looking for a way how to move part of `clearValue` method to the `Index Map`. However, it looked quite quirky and I've decided to remove it.

budnix commented on src/translations/maps/linkedPhysicalIndexToValueMap.js:
```diff
@@ -47,15 +48,16 @@ class LinkedPhysicalIndexToValueMap extends IndexMap {
    * @returns {boolean}
    */
-  setValueAtIndex(index, value) {
+  setValueAtIndex(index, value, position = this.orderOfIndexes.length) {
```
I haven't seen unit tests for the new argument `position`. Can I ask you to add it?

budnix replied:
Fixed within https://github.com/handsontable/handsontable/pull/7319/commits/9e403c163b67182122c4708f59abfaf3dd1efd11

budnix reviewed (changes requested):
~🤔  I've tried the use case https://github.com/handsontable/handsontable/pull/7319#issuecomment-712164184 using the newest changes but It seems that there is still a bug. Could you check that?~

I've checked again and I haven't found any issues.

budnix commented on src/plugins/filters/conditionUpdateObserver.js:
```diff
@@ -16,7 +16,7 @@ import { createArrayAssertion } from './utils';
  */
 class ConditionUpdateObserver {
-  constructor(conditionCollection, columnDataFactory = () => []) {
+  constructor(conditionCollection, columnDataFactory = () => [], hot) {
```
Fixed within https://github.com/handsontable/handsontable/pull/7319/commits/3356ad8640444fc0e564dda152d58350d6b1d92c

budnix commented:
While testing the "by_value" component I've found another bug where the content of the component is not refreshed after columns reorder.
![Kapture 2020-11-10 at 09 51 01](https://user-images.githubusercontent.com/571316/98651429-86b4c000-233a-11eb-9010-e4ff0bb90ce9.gif)

swistach reviewed (commented):
Shouldn't we restore possible values after remove column with active `filter_by_value`?
![altering_columns_filters](https://user-images.githubusercontent.com/11849621/99804408-3fe57800-2b3b-11eb-839d-3ebff6d30b03.gif)

budnix commented:
> Shouldn't we restore possible values after remove column with active `filter_by_value`?
> ![altering_columns_filters](https://user-images.githubusercontent.com/11849621/99804408-3fe57800-2b3b-11eb-839d-3ebff6d30b03.gif)

Filters are passive, so I'd say no. You can call the `hot.filter()` after column removal to achieve active-like behavior.

budnix commented on src/plugins/filters/conditionCollection.js:
```diff
@@ -1,37 +1,48 @@
      */
-    this.conditions = this.initConditionsCollection();
-
+    this.hot = hot;
```
Fixed within https://github.com/handsontable/handsontable/pull/7319/commits/f4cc1522265b42a7540858ae814ef51e35be358f

wojciechczerniak reviewed (approved):
🥇
